### PR TITLE
Create min/max functions that take bounds into account

### LINKED ISSUE #137: Create min/max functions that take bounds into account
It would be nice to be able to get the first and last item within the bounds an Interval to some precision.
Ex.
```
julia> function min(i::Interval{T, Open}; precision=eps(T)) where T
       return first(i) + precision
end 

julia> function min(i::Interval{T, Closed}; precision=eps(T)) where T
       return first(i)
end

julia> min(Interval{Open, Open}(1.0, 2.0); precision=0.1)
1.1

julia> min(Interval{Closed, Closed}(1.0, 2.0); precision=0.1)
1.0
```

### PR DESCRIPTION
**Pull request by Arvind-Maan**
fixes #137

### CODE REVIEW

File: src/interval.jl [hunk 1]
```diff
@@ -171,6 +171,22 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
     return R !== Unbounded ? interval.last : nothing
 end
 
+function Base.min(interval::Interval{T,L,R}; precision) where {T,L,R}
+    return Base.first(interval)
+end
+
+function Base.min(interval::Interval{T,Open,R}; precision) where {T,R}
```
Comment (Arvind-Maan) [inline src/interval.jl:178]: I deliberately didn't set precision as anything, since I feel like having a default could cause issues with different types. 

For example, [in the issue the example](https://github.com/invenia/Intervals.jl/issues/137#issue-678556586) used `eps(..)` as the default but this would cause errors on intervals that weren't floats. 

By leaving this keyword empty it makes the caller _have_ to define the type and how precise they want the function to be.

**Review by morris25 — APPROVED** [review]

File: src/interval.jl (hunk 1 above)
Comment (fchorney) [reply src/interval.jl:178]: If it's a keyword then it's not required to be called, so what would `max(I)` be? where `I` is some interval? 
I guess it would make sense to expect an error if you don't define the precision, but when the interval is closed you don't need to supply a precision, so it seems somewhat inconsistent in that way? 

That being said I'm not really sure how to solve that issue, as you are correct that different types would require different defaults.

Comment (Arvind-Maan) [reply src/interval.jl:178]: max(I) would cause an error since precision wouldn't be defined. 

One workaround I can think of is to define precision as None by default in the min/max functions that are `Unbounded` or `Closed` since precision doesn't matter in those.

This would make the call pass without changing the interface.

Comment (fchorney) [reply src/interval.jl:178]: I guess thinking about it more that should be fine. It makes sense that the function signatures be the same for the closed and open functions even if the precision does nothing on the closed ones. If we wanted viable defaults for precision we'd probably need to define the function for a bunch of different types which may not be worth it at this point.

**Review by fchorney — APPROVED** [review]

File: src/interval.jl [hunk 2]
```diff
@@ -171,6 +171,22 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
...
+    return interval.first + precision
+end
+
+function Base.max(interval::Interval{T,L,R}; precision=nothing) where {T,L,R}
+    return Base.last(interval)
+end
+
+function Base.max(interval::Interval{T,L,Open}; precision) where {T,L}
+    return interval.last - precision
```
Comment (omus) [inline src/interval.jl:187]:
```suggestion
    return last(interval) - precision
```

File: src/interval.jl (hunk 1 above)
Comment (omus) [reply src/interval.jl:178]: I would say that the default should be `eps` and for types where that value isn't defined it just needs to be passed in explicitly.

File: src/interval.jl [hunk 3]
```diff
@@ -171,6 +171,22 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
     return R !== Unbounded ? interval.last : nothing
 end
 
+function Base.min(interval::Interval{T,L,R}; precision=nothing) where {T,L,R}
```
[the remark this answers is not in the document]
Comment (Arvind-Maan) [reply src/interval.jl:174]: > Since the other `min` function doesn't set a default precision this function shouldn't either.

Since I added the default of eps(T) to the other function, I assume that it's fine to leave this default?

File: src/interval.jl [hunk 4]
```diff
@@ -171,6 +171,22 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
     return R !== Unbounded ? interval.last : nothing
 end
 
+function Base.minimum(interval::Interval{T,Closed,R}; precision=nothing) where {T,R}
```
Comment (omus) [inline src/interval.jl:174]: Just for consistency:
```suggestion
function Base.minimum(interval::Interval{T,Closed,R}; precision=eps(T)) where {T,R}
```

File: src/interval.jl [hunk 5]
```diff
@@ -171,6 +171,22 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
...
+    return first(interval)
+end
+
+function Base.minimum(interval::Interval{T,Open,R}; precision=eps(T)) where {T,R}
+    return first(interval) + precision
+end
+
+function Base.maximum(interval::Interval{T,L,Closed}; precision=nothing) where {T,L}
```
Comment (omus) [inline src/interval.jl:182]:
```suggestion
function Base.maximum(interval::Interval{T,L,Closed}; precision=eps(T)) where {T,L}
```

File: src/interval.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (omus) [reply src/interval.jl]: I'd keep the two methods consistent

File: src/interval.jl [hunk 6]
```diff
@@ -171,6 +171,22 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
...
 end
 
+function Base.minimum(interval::Interval{T,Closed,R}; precision=nothing) where {T,R}
+    return first(interval)
+end
+
+function Base.minimum(interval::Interval{T,Open,R}; precision=eps(T)) where {T,R}
+    return first(interval) + precision
+end
+
```
Comment (omus) [inline src/interval.jl:181]: You can make things nicer for integers by doing:
```suggestion

function Base.minimum(interval::Interval{T,L,R}; precision=one(T)) where {T <: Integer, L, R}
    return minimum(interval; precision=precision)
end

```

Comment (omus) [reply src/interval.jl:181]: We should also probably special case `AbstractFloat` as with very large floating-point values adding `eps` is inaccurate:
```julia
julia> x = nextfloat(typemin(Float64))
-1.7976931348623157e308

julia> x + eps(Float64)
-1.7976931348623157e308

julia> x + eps(Float64) == x
true

julia> nextfloat(x)
-1.7976931348623155e308

julia> nextfloat(x) == x
false
```

**Comment by omus** [comment]
Another thing I'm thinking about is the name "precision". Currently Julia already has a `precision` function which returns an integer. I think the term `increment` may be better suited

File: src/interval.jl (hunk 4 above)
Comment (omus) [reply src/interval.jl:174]: Actually, I take this back. You can leave it as `nothing`. I thought Julia was smart enough to optimize this away. That isn't the case.
```julia
julia> f(x=nothing) = 1
f (generic function with 2 methods)

julia> g(x=eps(Int)) = 1
g (generic function with 2 methods)

julia> @code_native f()
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ REPL[5]:1 within `f'
	movl	$1, %eax
	retq
	nopw	%cs:(%rax,%rax)
; └

julia> @code_native g()
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ REPL[6]:1 within `g'
	pushq	%rax
	movabsq	$jl_system_image_data, %rax
	movq	%rax, (%rsp)
	movabsq	$jl_apply_generic, %rax
	movabsq	$jl_system_image_data, %rdi
	movq	%rsp, %rsi
	movl	$1, %edx
	callq	*%rax
	ud2
	nop
; └

julia> g()
ERROR: MethodError: no method matching eps(::Type{Int64})
Closest candidates are:
  eps(::Dates.Time) at /Users/omus/Development/Julia/1.3/usr/share/julia/stdlib/v1.3/Dates/src/types.jl:387
  eps(::Dates.Date) at /Users/omus/Development/Julia/1.3/usr/share/julia/stdlib/v1.3/Dates/src/types.jl:386
  eps(::Dates.DateTime) at /Users/omus/Development/Julia/1.3/usr/share/julia/stdlib/v1.3/Dates/src/types.jl:385
  ...
Stacktrace:
 [1] g() at ./REPL[6]:1
 [2] top-level scope at REPL[9]:1
```

File: src/interval.jl (hunk 5 above)
Comment (omus) [reply src/interval.jl:182]: See https://github.com/invenia/Intervals.jl/pull/141/files#r488735402

File: src/interval.jl [hunk 7]
```diff
@@ -171,6 +171,38 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
...
+function Base.minimum(interval::Interval{T,Open,R}; increment=eps(T)) where {T<:AbstractFloat,R}
+    return nextfloat(first(interval))
+end
+
+function Base.maximum(interval::Interval{T,L,Open}; increment=eps(T)) where {T<:AbstractFloat,L}
+    return prevfloat(last(interval))
+end
+
```
Comment (omus) [inline src/interval.jl:205]: Can we move these definitions to after the `isbounded` function definition? Also, these functions should probably use `AbstractInterval` instead of `Interval` so that this also works for `AnchoredIntervals`.

File: test/interval.jl [hunk 8]
```diff
@@ -125,12 +125,33 @@ isinf(::TimeType) = false
...
+                # the value we compare to min/max depends on if the bound is open/closed
+                # for floats, adding/subtracting eps(Float) is inaccurate for large vals
+                # so we must use next/prev float
+
+                min_comp = if a in interval
+                    first(interval)
+                else
+                    typeof(first(interval)) <: AbstractFloat ?
+                        nextfloat(first(interval)) :
+                        first(interval) + p
```
Comment (omus) [inline test/interval.jl:143]:
```suggestion
                    val = first(interval)
                    eltype(interval) isa AbstractFloat ? nextfloat(val) : val + p
```

File: test/interval.jl [hunk 9]
```diff
@@ -125,12 +125,33 @@ isinf(::TimeType) = false
...
             for (L, R) in BOUND_PERMUTATIONS
                 interval = Interval{L, R}(a, b)
 
                 @test first(interval) == a
                 @test last(interval) == b
+                # the value we compare to min/max depends on if the bound is open/closed
+                # for floats, adding/subtracting eps(Float) is inaccurate for large vals
+                # so we must use next/prev float
+
+                min_comp = if a in interval
```
Comment (omus) [inline test/interval.jl:138]:
```suggestion

                # The value we compare to min/max depends on if the bound is open/closed
                # for floats, adding/subtracting eps(Float) is inaccurate for large vals
                # so we must use next/prev float
                min_comp = if a in interval
```

File: src/interval.jl (hunk 7 above)
Comment (omus) [reply src/interval.jl:205]: We should also have tests for `AnchoredInterval`s and intervals using `Unbounded`

File: src/interval.jl [hunk 10]
```diff
@@ -171,6 +171,38 @@ function Base.last(interval::Interval{T,L,R}) where {T,L,R}
...
+    return last(interval) - increment
+end
+
+function Base.minimum(interval::Interval{T,Open,R}; increment=eps(T)) where {T<:AbstractFloat,R}
+    return nextfloat(first(interval))
+end
+
+function Base.maximum(interval::Interval{T,L,Open}; increment=eps(T)) where {T<:AbstractFloat,L}
+    return prevfloat(last(interval))
+end
```
Comment (omus) [inline src/interval.jl:204]:
```suggestion
function Base.minimum(interval::Interval{T,Open,R}; increment=nothing) where {T<:AbstractFloat,R}
    min_val = first(interval)
    return increment !== nothing ? min_val + increment : nextfloat(min_val)
end

function Base.maximum(interval::Interval{T,L,Open}; increment=nothing) where {T<:AbstractFloat,L}
    max_val = last(interval)
    return increment !== nothing ? max_val - increment : prevfloat(max_val)
end
```

File: test/interval.jl (hunk 8 above)
Comment (omus) [reply test/interval.jl:143]: Actually, this extra logic can be removed if `AbstractFloat` respects explicitly set increment values

File: test/interval.jl [hunk 11]
```diff
@@ -125,12 +125,43 @@ isinf(::TimeType) = false
...
+        for (a, b, p) in append!(accessor_specific_vals, test_values)
             for (L, R) in BOUND_PERMUTATIONS
                 interval = Interval{L, R}(a, b)
 
                 @test first(interval) == a
                 @test last(interval) == b
+
+                # If p=eps(Float64) then we treat it as if p=nothing.
+                # This is because adding eps to floats is not accurate and we must use nextfloat and prevfloat.
+                p = p == eps() ? nothing : p
```
Comment (omus) [inline test/interval.jl:145]: Doesn't this result in `(-10.0, 10.0, nothing)` and `(-10.0, 10.0, eps(Float64))` being identical tests?

Comment (Arvind-Maan) [reply test/interval.jl]: Whooops, this is old code that I accidentally pushed. Will resolve.

File: src/interval.jl [hunk 12]
```diff
@@ -187,6 +187,47 @@ Base.isopen(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Open && R =
...
+
+# Minimum and maximum calls for open bounds and ambigous types, return open side +- increment
+# Default increment val is eps(T)
+function Base.minimum(interval::AbstractInterval{T,Open,R}; increment=eps(T)) where {T,R}
+    return first(interval) + increment
+end
+
+function Base.maximum(interval::AbstractInterval{T,L,Open}; increment=eps(T)) where {T,L}
+    return last(interval) - increment
+end
```
Comment (Arvind-Maan) [inline src/interval.jl:207]: What should I do in the following case:

``` julia
first(interval) + increment > last(interval)
last(interval) - increment < first(interval)
```

should I be returning `nothing` in these cases? Since they are ∉ the interval

File: test/interval.jl [hunk 13]
```diff
@@ -149,6 +177,58 @@ isinf(::TimeType) = false
...
+            (nothing, nothing, 1, typeof(1), Unbounded, Unbounded),
+            (nothing, nothing, 0.01, Float64, Unbounded, Unbounded),
+            (nothing, nothing, 1, Char, Unbounded, Unbounded),
+            (nothing, nothing, Day(1), Day, Unbounded, Unbounded),
+            (nothing, nothing, Millisecond(1), DateTime, Unbounded, Unbounded),
+
+            # test adding eps() with unbounded
+            (-10.0, nothing, nothing, Float64, Open, Unbounded),
+            (nothing, 10.0, eps(Float64), Float64, Unbounded, Open),
+        ]
```
Comment (omus) [inline test/interval.jl:199]: May as well construct actual intervals in this case as you have all of the required information:
```suggestion
        unbounded_test_values = [
            # one side unbounded with different types
            (Interval{Open,Unbounded}(-10, nothing), 1)
            (Interval{Unbounded,Closed}(nothing, 1.0), 0.01),
            (Interval{Unbounded,Open}(nothing, 'z') 1),
            (Interval{Closed,Unbounded}(Date(2013, 2, 13), nothing), Day(1)),
            (Interval{Open,Unbounded}(DateTime(2016, 8, 11, 0, 30), nothing), Millisecond(1)),

            # both sides unbounded different types
            (Interval{Int}(nothing, nothing), 1),
            (Interval{Float64}(nothing, nothing), 0.01),
            (Interval{Char}(nothing, nothing), 1),
            (Interval{Day}(nothing, nothing), Day(1)),
            (Interval{DateTime}(nothing, nothing), Millisecond(1)),

            # test adding eps() with unbounded
            (Interval{Open,Unbounded}(-10.0, nothing), nothing),
            (Interval{Unbounded,Open}(nothing, 10.0), eps(Float64)),
        ]
```

File: test/interval.jl [hunk 14]
```diff
@@ -124,19 +124,47 @@ isinf(::TimeType) = false
...
+
+                ma = if b ∈ interval
+                    last(interval)
+                else
+                    v = last(interval)
+                    unit === nothing && eltype(interval) <: AbstractFloat ? prevfloat(v) : v - unit
+                end
+
+                @test minimum(interval; increment=unit) == mi
+                @test maximum(interval; increment=unit) == ma
```
Comment (omus) [inline test/interval.jl:160]: I think I'd combine these test with what you have below and call the testset "minimum/maximum"

File: test/interval.jl [hunk 15]
```diff
@@ -149,6 +177,58 @@ isinf(::TimeType) = false
...
+        for (a, b, unit, T, L, R) in unbounded_test_values
+            interval = Interval{T,L,R}(a, b)
+
+            @test first(interval) == a
+            @test last(interval) == b
+
+            # The value we compare to min/max depends on if the bound is open/closed.
+            # we can determine if a bound is open/closed by checking if the edge in the interval
+            mi = if a === nothing
+                nothing
```
Comment (omus) [inline test/interval.jl:209]: I was originally thinking when using `minimum` on an left-unbounded interval you would get the minimum possible value in that interval. When taking into account the element type this could return `typemin(eltype(interval))`. There can definitely be some discussion on what we expect in this case

File: src/interval.jl (hunk 12 above)
Comment (omus) [reply src/interval.jl:207]: A very good question. As the provided increment is supposed to be the smallest span between to values I was thinking an exception would be reasonable. Looking at empty intervals (e.g. `Interval{Open,Open}(0, 0)`) should probably be similar to using `minimum`/`maximum` on an empty collection which raises an exception:
```julia
julia> maximum(Int[])
ERROR: ArgumentError: reducing over an empty collection is not allowed
```

File: test/anchoredinterval.jl [hunk 16]
```diff
@@ -187,6 +191,8 @@ using Intervals: Bounded, Ending, Beginning, canonicalize, isunbounded
         interval = AnchoredInterval{Day(1)}(startpoint)
         @test first(interval) == startpoint
         @test last(interval) == ZonedDateTime(2018, 3, 12, tz"America/Winnipeg")
+        @test minimum(interval) == startpoint
+        @test maximum(interval, increment=Hour(1)) == ZonedDateTime(2018, 3, 11, 23, tz"America/Winnipeg")
```
[the remark this answers is not in the document]
Comment (Arvind-Maan) [reply test/anchoredinterval.jl:195]: eps(ZonedDateTime) creates an error unfortunately.

File: src/interval.jl (hunk 12 above)
Comment (morris25) [reply src/interval.jl:207]: I suppose it would depend if the other side is open or closed. You would expect an `Interval{T, Closed, Open}` to have a maximum regardless of increment.

File: test/interval.jl [hunk 17]
```diff
@@ -144,6 +143,164 @@
...
+                l - unit
+            end
+
+            nv ∈ interval && return nv
+
+            # If we get to this point, the min/max functions throw a DomainError
+            # Since we want our tests to be predictable, we will not throw an error in this helper.
+        end
+        @testset "bounded intervals" begin
+            bounded_test_vals = [
```
Comment (Arvind-Maan) [inline test/interval.jl:208]: Can't use Infinity.jl since `isfinite`does ` ∞ -  ∞` which causes an error. Could make a workaround for this

**Comment by nickrobinson251** [comment]
As a follow-up to this PR, it'd be nice to resolve https://github.com/invenia/Intervals.jl/issues/90 😊

File: src/interval.jl [hunk 18]
```diff
@@ -187,6 +187,72 @@ Base.isopen(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Open && R =
 isunbounded(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Unbounded && R === Unbounded
 isbounded(interval::AbstractInterval{T,L,R}) where {T,L,R} = L !== Unbounded && R !== Unbounded
 
+# Minimum and maximum calls for Closed and Unbounded Intervals, just return first and last.
+function Base.minimum(interval::AbstractInterval{T,L,R}; increment=nothing) where {T,L,R}
+    first(interval) === nothing && return typemin(T)
+    return first(interval)
+end
```
Comment (omus) [inline src/interval.jl:194]:
```suggestion
function Base.minimum(interval::AbstractInterval{T,L,R}; increment=nothing) where {T,L,R}
    return L === Unbounded ? typemin(T) : interval.first
end
```
OR use multiple dispatch to remove the conditional check:
```suggestion
Base.minimum(interval::AbstractInterval{T,Unbounded,R}; increment=nothing) where {T,R}
    return typemin(T)
end

Base.minimum(interval::AbstractInterval{T,Closed,R}; increment=nothing) where {T,R}
    return interval.first
end
```
Note: I'm using `interval.first` over `first(interval)` here to avoid the unbounded check `first` does

File: src/interval.jl [hunk 19]
```diff
@@ -187,6 +187,72 @@ Base.isopen(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Open && R =
...
+
+function Base.maximum(interval::AbstractInterval{T,L,R}; increment=nothing) where {T,L,R}
+    last(interval) === nothing && return typemax(T)
+    return last(interval)
+end
+
+# Minimum and maximum calls for open bounds and ambigous types, return open side +- increment
+# Default increment val is eps(T)
+function Base.minimum(interval::AbstractInterval{T,Open,R}; increment=eps(T)) where {T,R}
+    isempty(interval) && return nothing
```
Comment (omus) [inline src/interval.jl:204]: I think this should error

File: src/interval.jl [hunk 20]
```diff
@@ -187,6 +187,72 @@ Base.isopen(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Open && R =
...
+end
+
+function Base.maximum(interval::AbstractInterval{T,L,R}; increment=nothing) where {T,L,R}
+    last(interval) === nothing && return typemax(T)
+    return last(interval)
+end
+
+# Minimum and maximum calls for open bounds and ambigous types, return open side +- increment
+# Default increment val is eps(T)
+function Base.minimum(interval::AbstractInterval{T,Open,R}; increment=eps(T)) where {T,R}
```
Comment (omus) [inline src/interval.jl:203]: I'd group the `minimum` functions all together

File: src/interval.jl [hunk 21]
```diff
@@ -187,6 +187,72 @@ Base.isopen(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Open && R =
...
+    return maximum(interval, increment=one(T))
+end
+
+# Minimum and maximum calls for open bounds and Float types, return open side +- increment
+# Default increment val for Floats is nothing, which returns nextfloat or prevfloat (i.e., +- eps())
+function Base.minimum(interval::AbstractInterval{T,Open,R}; increment=nothing) where {T<:AbstractFloat,R}
+    isempty(interval) && return nothing
+    min_val = first(interval)
+    # Since intervals can't have NaN, we can just use !isfinite to check if infinite
+    next_val = if !isfinite(min_val) || increment === nothing || increment == eps(T)
```
Comment (omus) [inline src/interval.jl:235]: I'd only use `nextfloat` if a increment isn't explicitly given. Even if it is `eps(T)`
```suggestion
    next_val = if !isfinite(min_val) || increment === nothing
```

File: test/anchoredinterval.jl [hunk 22]
```diff
@@ -175,6 +177,9 @@ using Intervals: Bounded, Ending, Beginning, canonicalize, isunbounded
 
         @test first(interval) == Date(2016, 8, 11)
         @test last(interval) == Date(2016, 8, 12)
+        # throws domain error
+        @test_throws DomainError minimum(interval, increment=Day(1)) == first(interval) + Day(1)
+        @test_throws DomainError maximum(interval, increment=Day(1)) == last(interval) - Day(1)
```
Comment (omus) [inline test/anchoredinterval.jl:182]:
```suggestion
        @test_throws DomainError minimum(interval, increment=Day(1))
        @test_throws DomainError maximum(interval, increment=Day(1))
```

File: test/anchoredinterval.jl (hunk 16 above)
[the remark this answers is not in the document]
Comment (omus) [reply test/anchoredinterval.jl:196]: We should be using the smallest value supported by the element type as that's what `increment` is intended to be set to:
```suggestion
        @test maximum(interval, increment=Millisecond(1)) == last(interval) - Millisecond(1)
```

File: test/anchoredinterval.jl [hunk 23]
```diff
@@ -197,24 +204,32 @@ using Intervals: Bounded, Ending, Beginning, canonicalize, isunbounded
         interval = AnchoredInterval{Day(1)}(startpoint)
         @test first(interval) == startpoint
         @test last(interval) == ZonedDateTime(2018, 11, 5, tz"America/Winnipeg")
+        @test minimum(interval) == startpoint
+        @test maximum(interval, increment=Hour(1)) == ZonedDateTime(2018, 11, 4, 23, tz"America/Winnipeg")
```
Comment (omus) [inline test/anchoredinterval.jl:208]:
```suggestion
        @test maximum(interval, increment=Millisecond(1)) == last(interval) - Millisecond(1)
```

File: test/anchoredinterval.jl [hunk 24]
```diff
@@ -197,24 +204,32 @@ using Intervals: Bounded, Ending, Beginning, canonicalize, isunbounded
...
         @test span(interval) == Day(1)
 
         endpoint = ZonedDateTime(2020, 3, 9, 2, tz"America/Winnipeg")
         interval = AnchoredInterval{Day(-1)}(endpoint)
         @test_throws NonExistentTimeError first(interval)
+        @test_throws NonExistentTimeError minimum(interval, increment=Hour(1))
         @test last(interval) == endpoint
+        @test maximum(interval) == endpoint
```
Comment (omus) [inline test/anchoredinterval.jl:216]: Be consistent with ordering
```suggestion
        @test_throws NonExistentTimeError first(interval)
        @test last(interval) == endpoint
        @test_throws NonExistentTimeError minimum(interval, increment=Millisecond(1))
        @test maximum(interval) == endpoint
```

File: test/interval.jl (hunk 17 above)
Comment (omus) [reply test/interval.jl:208]: If you rebase your changes this should no longer be an issue:
```julia
julia> isfinite(∞)
false

julia> Intervals.isfinite(∞)
false
```

Comment (Arvind-Maan) [reply test/interval.jl:208]: Might need some more clarification on this.

I rebased in the traditional sense `git rebase master` and I also updated from master, but it still doesn't work for some reason. 🙁 

If I replicate your calls in the REPL, it works fine as well.

Comment (Arvind-Maan) [reply test/interval.jl:208]:
```julia
julia> using Dates,Infinity, Intervals

julia> x = Interval{Open, Open}(Date(2013, 2, 13), ∞)
Interval{InfExtendedTime{Date},Open,Open}(InfExtendedTime{Date}(Date("2013-02-13")), InfExtendedTime{Date}(∞))

julia> minimum(x, increment=Day(1))
InfExtendedTime{Date}(Date("2013-02-14"))

julia> maximum(x, increment=Day(1))
ERROR: ∞-∞ is undefined
Stacktrace:
 [1] + at /Users/arvindmaan/.julia/packages/Infinity/ezs8D/src/infinite/arithmetic.jl:9 [inlined]
 [2] -(::Infinite, ::Infinite) at /Users/arvindmaan/.julia/packages/Infinity/ezs8D/src/infinite/arithmetic.jl:10
 [3] -(::InfExtendedTime{Date}, ::InfExtendedTime{Date}) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Dates/src/arithmetic.jl:9
 [4] isfinite(::InfExtendedTime{Date}) at /Users/arvindmaan/.julia/dev/Intervals/src/isfinite.jl:3
 [5] maximum(::Interval{InfExtendedTime{Date},Open,Open}; increment::Day) at /Users/arvindmaan/.julia/dev/Intervals/src/interval.jl:228
 [6] top-level scope at REPL[7]:1
```

File: src/docstrings.jl [hunk 25]
```diff
@@ -105,3 +105,31 @@ Note using `!isbounded` is commonly used to determine if any end of the interval
...
+
+"""
+    minimum(interval::AbstractInterval{T}; increment::Any) -> T
+
+The minimum value in the interval.
+
+If left bound is closed, returns `first(interval)`.
+If left bound is unbounded, returns `typemin(T)`
+
+On an empty interval or when the increment results in `first(interval) + increment ∉ interval`, returns BoundsError.
```
Comment (omus) [inline src/docstrings.jl:117]:
```suggestion
A `BoundsError` is thrown for empty intervals or when the increment results in a minimum value not-contained by the interval.
```

File: src/docstrings.jl [hunk 26]
```diff
@@ -105,3 +105,31 @@ Note using `!isbounded` is commonly used to determine if any end of the interval
...
+
+When T <: AbstractFloat and no increment is passed, defaults to returning `nextfloat(first(interval))`
```
Comment (omus) [inline src/docstrings.jl:119]: I think we can leave this implementation detail out.

File: src/interval.jl [hunk 27]
```diff
@@ -187,70 +187,63 @@ Base.isopen(interval::AbstractInterval{T,L,R}) where {T,L,R} = L === Open && R =
...
-
-# Minimum and maximum calls for open bounds and ambigous types, return open side +- increment
-# Default increment val is eps(T)
 function Base.minimum(interval::AbstractInterval{T,Open,R}; increment=eps(T)) where {T,R}
-    isempty(interval) && return nothing
+    isempty(interval) && throw(BoundsError(interval, 0))
     min_val = first(interval)
+    # Since intervals can't have NaN, we can just use !isfinite to check if infinite
     !isfinite(min_val) && return typemin(T)
     min_val + increment ∈ interval && return min_val + increment
```
Comment (omus) [inline src/interval.jl]:
```suggestion
    min_val = interval.first + increment
    # Since intervals can't have NaN, we can just use !isfinite to check if infinite
    !isfinite(min_val) && return typemin(T)
    min_val ∈ interval && return min_val
```

File: test/interval.jl (hunk 17 above)
Comment (omus) [reply test/interval.jl:208]: The trouble is definitely from `InfExtendedTime` specifically here. I'll see if I can come up with a solution for that

Comment (omus) [reply test/interval.jl:208]: Add the following to the end of `src/isfinite.jl` and everything should be working:
```julia
isfinite(x::Union{Type{T}, T}) where T<:TimeType = Base.isfinite(x)
```

**Comment by omus** [comment]
Thanks for sticking with this. Things are shaping up nicely now :)

File: test/anchoredinterval.jl [hunk 28]
```diff
@@ -175,6 +177,9 @@ using Intervals: Bounded, Ending, Beginning, canonicalize, isunbounded
 
         @test first(interval) == Date(2016, 8, 11)
         @test last(interval) == Date(2016, 8, 12)
+        # throws domain error
```
Comment (omus) [inline test/anchoredinterval.jl:180]:
```suggestion
```

File: test/interval.jl [hunk 29]
```diff
@@ -144,6 +143,141 @@
...
+                prevfloat(l)
+            else
+                l - unit
+            end
+
+            nv ∈ interval && return nv
+
+            # If we get to this point, the min/max functions throw a DomainError
+            # Since we want our tests to be predictable, we will not throw an error in this helper.
+        end
```
Comment (omus) [inline test/interval.jl:206]: As theses helpers pretty much just re-implement minimum/maximum there aren't that helpful with testing. I think we can proceed with this but something to keep in mind for future work.

File: src/docstrings.jl [hunk 30]
```diff
@@ -105,3 +105,29 @@ Note using `!isbounded` is commonly used to determine if any end of the interval
 unbounded.
 """
 isbounded(::AbstractInterval)
+
+"""
+    minimum(interval::AbstractInterval{T}; [increment]) -> T
+
+The minimum value in the interval contained within the interval.
```
Comment (omus) [inline src/docstrings.jl:112]:
```suggestion
    minimum(interval::AbstractInterval{T}; [increment]) -> T

The minimum value contained within the `interval`.
```

File: src/docstrings.jl [hunk 31]
```diff
@@ -105,3 +105,29 @@ Note using `!isbounded` is commonly used to determine if any end of the interval
...
+
+A `BoundsError` is thrown for empty intervals or when the increment results in a minimum value
+not-contained by the interval.
+"""
+minimum(::AbstractInterval; increment)
+
+"""
+    maximum(interval::AbstractInterval{T}; increment::Any) -> T
+
+The maximum value in the interval contained within the interval.
```
Comment (omus) [inline src/docstrings.jl:125]:
```suggestion
    maximum(interval::AbstractInterval{T}; [increment]) -> T

The maximum value contained within the `interval`.
```

File: src/docstrings.jl [hunk 32]
```diff
@@ -105,3 +105,29 @@ Note using `!isbounded` is commonly used to determine if any end of the interval
 unbounded.
 """
 isbounded(::AbstractInterval)
+
+"""
+    minimum(interval::AbstractInterval{T}; [increment]) -> T
+
+The minimum value in the interval contained within the interval.
+
+If left-closed, returns `first(interval)`.
```
Comment (omus) [inline src/docstrings.jl:114]:
```suggestion
If left-closed, returns `first(interval)`.
If left-open, returns `first(interval) + eps(first(interval))`
```

File: src/docstrings.jl [hunk 33]
```diff
@@ -105,3 +105,29 @@ Note using `!isbounded` is commonly used to determine if any end of the interval
...
+not-contained by the interval.
+"""
+minimum(::AbstractInterval; increment)
+
+"""
+    maximum(interval::AbstractInterval{T}; increment::Any) -> T
+
+The maximum value in the interval contained within the interval.
+
+If right-closed, returns `last(interval)`.
```
Comment (omus) [inline src/docstrings.jl:127]:
```suggestion
If right-closed, returns `last(interval)`.
If right-open, returns `last(interval) - eps(last(interval))`
```

**Review by omus — APPROVED** [review]
With these last changes we should be good.

File: test/interval.jl (hunk 29 above)
Comment (Arvind-Maan) [reply test/interval.jl:206]: I agree, in hindsight it would've been best to separate the tests per edge case and avoid having to do this.

File: src/docstrings.jl (hunk 30 above)
Comment (omus) [reply src/docstrings.jl]: Looks like this change was missed?

File: src/docstrings.jl (hunk 31 above)
Comment (omus) [reply src/docstrings.jl]: Looks like this change was missed?

**Comment by omus** [comment]
PR was meant to be squashed. Squashed commit is https://github.com/invenia/Intervals.jl/commit/c37cbb77cab32c842cd482ce3170eeb18b64d9e9
